### PR TITLE
`metamath-extract` can now remove lemmas already present in another database

### PR DESCRIPTION
**Pull request by nishantjr**
Parallels https://github.com/runtimeverification/proof-generation/pull/77

### CODE REVIEW

File: bin/metamath-compress [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
+# Modified from https://github.com/metamath/set.mm/blob/develop/scripts/verify
+
+set -euo pipefail
+
+# We should find a cleaner way of doing this long term.
```
Comment (fiedlr) [inline bin/metamath-compress]: How about we merge this with how I implemented it in `qed.sh`? My version looks more compact.

File: scripts/metamath-extract.py [hunk 2]
```diff
@@ -214,7 +217,7 @@ def abbreviate(name: str) -> str:
 
 def abbreviate_constructors(term: Term) -> Term:
     if isinstance(term, Metavariable):
-        return term
+        return Metavariable(abbreviate(term.name))
```
Comment (fiedlr) [inline scripts/metamath-extract.py:218]: What does this have to do with  this PR? It should ideally be in another commit, right?

File: scripts/metamath-extract.py [hunk 3]
```diff
@@ -232,46 +235,98 @@ def abbreviate_lemmas(statement: Statement) -> Statement:
...
-                lemmas, lemma_applications = deconstruct_compressed_proof(proof)
+                lemmas, lemma_applications = deconstruct_compressed_proof(statement)
                 lemmas = (abbreviate(lemma) for lemma in lemmas)
                 proof = '( {} ) {}'.format(' '.join(lemmas), lemma_applications)
             return ProvableStatement(new_label, terms_abbreviate_constructors(statement.terms), proof=proof)
         else:
             raise RuntimeError(f'Unexpected statement type, {type(statement)}')
 
 
+def collect_theorem_names(database: Database) -> set[str]:
```
Comment (fiedlr) [inline scripts/metamath-extract.py]: `collect_provable_names`?

File: scripts/metamath-extract.py [hunk 4]
```diff
@@ -175,10 +177,11 @@ def slice_database(input_database: Database) -> Iterator[tuple[str, Database]]:
             cut_antecedents[axiom_get_label(statement)] = cast('AxiomaticStatement | Block', statement)
         elif isinstance(statement, (ProvableStatement, Block)):
             antecedents, consequent = deconstruct_provable(statement)
-            yield (
-                consequent.label,
-                supporting_database_for_provable(cut_antecedents, global_disjoints, consequent, antecedents),
-            )
+            if (include and consequent.label in include) and (consequent.label not in exclude):
```
Comment (fiedlr) [inline scripts/metamath-extract.py]: Should this not be 
```
if (not include or consequent.label in include) and (consequent.label not in exclude):
```
? I would assume that having an undefined `include` means that this option is not used, and thus everything is included.

File: scripts/metamath-extract.py [hunk 5]
```diff
@@ -232,46 +235,98 @@ def abbreviate_lemmas(statement: Statement) -> Statement:
...
 
+    exclude: set[str] = set()
+    if args.exclude:
+        print('Parsing exclude database...', end='', flush=True)
+        excluded_database = load_database(args.exclude, include_proof=False)
+        print(' Done.')
+        exclude = collect_theorem_names(excluded_database)
+
+    deps = dependency_graph(input_database)
+    include: set[str] | None = transitive_closure(deps, ['goal'])
```
Comment (fiedlr) [inline scripts/metamath-extract.py]: How can this be `None`?

File: scripts/metamath-extract.py (hunk 2 above)
Comment (nishantjr) [reply scripts/metamath-extract.py:218]: I did put it in a standalone commit, though I wish we could merge the PR without squashing commits. This PR needs this fix because, the new changes  do not verify without it. 

https://github.com/fiedlr/k-proof-generation/pull/17/commits/1453bcbfb3f0619ea1f25d7048d6b0e6e7f9a722

File: scripts/metamath-extract.py (hunk 5 above)
Comment (nishantjr) [reply scripts/metamath-extract.py]: The original idea was to have the main goal specified on the command line, but we're not doing that now. I'll remove this case.

**Comment by fiedlr** [comment]
LGTM

**Review by fiedlr — APPROVED** [review]
